### PR TITLE
Feature: Refactor Lesson Table of Contents to a Stimulus JS

### PR DESCRIPTION
**Pull request by KevinMulhern**
Because:
* Our old table of contents behaviour had a dependency on ScrollSpy - a Bootstrap utility.
* We want to move all of our JavaScript out of scripts and into Stimulus JS controllers.

This commit:
* Change style of table of contents.
  * Title to make it clear what the sidebar contains
  * Left border to separate it from lesson content
* Active sections are highlighted with gray background and left border color darken.
* Add lesson table of contents Stimulus controller which:
  * Builds the table of contents
  * Handles the active scrolling sections using intersection observer.
* Show all top level h3 headings from the lesson content in the table of contents instead of filtering and displaying only the common lesson headings.

Complete the following REQUIRED checkboxes:

-   [x] I have thoroughly read and understand [The Odin Project Contributing Guide](https://github.com/TheOdinProject/theodinproject/blob/main/CONTRIBUTING.md)
-   [x] The title of this PR follows the `keyword: brief description of change` format, using one of the following keywords:
  - `Feature` - adds new or amends existing user-facing behaviour
  - `Chore` - changes that have no user-facing value, refactors, dependency bumps, etc
  - `Fix` - bug fixes

Complete the following checkbox ONLY IF it is applicable to your PR. You can complete it later if it is not currently applicable:
-   [x] I have verified all tests and linters pass against my changes, and/or I have included automated tests where applicable

<hr>

Screenshot 2022-08-05 at 08 21 35

### CODE REVIEW

File: app/javascript/stylesheets/typography.css [hunk 1]
```diff
@@ -10,4 +10,8 @@
   p {
     @apply text-gray-500 text-lg tracking-wide;
   }
+
+  .toc-item-active {
```
Comment (KevinMulhern) [inline app/javascript/stylesheets/typography.css:14]: Add to lesson-content.css after https://github.com/TheOdinProject/theodinproject/pull/3137 is merged

File: app/views/lessons/show.html.erb [hunk 2]
```diff
@@ -4,15 +4,18 @@
...
+      <div class="lesson-content max-w-prose mx-auto xl:mx-0 text-lg" data-controller="syntax-highlighting" data-lesson-toc-target="lessonContent">
         <%= @lesson.content.html_safe %>
       </div>
     </article>
 
-     <aside class="col-span-4 col-start-10 hidden xl:block">
-      <div class="sticky top-12 pb-20 lesson-navigation"></div>
+    <aside class="col-span-4 col-start-10 hidden xl:block">
+      <div class="sticky top-12 pb-20">
+        <h4 class="text-xl font-semibold pb-4 text-gray-500 odin-dark-light-gray-font">In this lesson</h4>
```
Comment (KevinMulhern) [inline app/views/lessons/show.html.erb:16]: update colors to match new lesson content colors after https://github.com/TheOdinProject/theodinproject/pull/3137 is merged

File: app/assets/stylesheets/components/lesson/lesson_content.scss [hunk 3]
```diff
@@ -1,5 +1,9 @@
 .lesson-content {
 
+  section[id] {
```
Comment (KevinMulhern) [inline app/assets/stylesheets/components/lesson/lesson_content.scss:3]: Add this to lesson-content.css file after https://github.com/TheOdinProject/theodinproject/pull/3137 is merged

File: app/javascript/controllers/lesson_toc_controller.js [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    this.buildTocItems();
+
+    this.lessonContentTarget.querySelectorAll('section[id]').forEach((section) => {
+      this.tocItemObserver().observe(section);
+    });
+  }
+
+  tocItemObserver() {
+    return (
+      new IntersectionObserver((entries) => {
```
Comment (ChargrilledChook) [inline app/javascript/controllers/lesson_toc_controller.js:21]: TIL about `IntersectionObserver`, neat

**Review by ChargrilledChook — APPROVED** [review]
This looks absolutely brilliant mate :fire: :fire: :fire: 

Love the new style, feels much more dynamic. (Side note, was Scrollspy supposed to have a similar highlighting effect? I never noticed the TOC changing styles before this)

Having a proper dynamic TOC makes the whole thing much more useful, too.

Awesome stuff :rocket:

**Review by leosoaivan — APPROVED** [review]

**Comment by KevinMulhern** [comment]
> Side note, was Scrollspy supposed to have a similar highlighting effect?
Yea it was 🙈  I must of broke it with some of the recent changes. But nobody noticed it wasn't working lol 😆

**Pull request merged**